### PR TITLE
8292697: ZGC: Hangs when almost out of metaspace memory

### PR DESCRIPTION
**Pull request by stefank**
HotSpot performs "critical metaspace allocations" when it's running out of metaspace. The failed allocations are registered in a queue, which the GC prioritizes when cleaning up the metaspace. There's a race in the code that handles these requests.

These requests are added to the queue, and the GC will "process" each request in turn when it runs the metaspace purge phase. The queue handling has an optimization that says that only the first request in the queue needs to trigger the GC, all subsequent requests in the queue will wait for that GC. When the GC gets to the purge phase it will mark in all requests that they have been processed. Note: that this doesn't mean that the request was satisfied, it could be that the result was NULL (and the thread will trigger a last-effort GC before it throws an OOME).

The bug is in the code that determines if a request is responsible for triggering a new GC. The current code just checks if the current request is first in the queue. This doesn't work if the code is called just after the GC has run purge, but before the old requests have been removed. The new request sees that there are already elements in the queue, so it doesn't trigger the GC. And at the same time, the old requests have been processed and they won't trigger the GC either. So, now the new request is waiting for a GC that will not be triggered by anyone.

Note: The reason why there's a delay between the GC processing a request, and the removal from the queue, is that the Java thread that added the request is also responsible for removing the request form the queue. The reason for this is that the mentioned last-effort GC, needs to be able to process the request a second time.

The fix-proposal is to let threads adding new requests check if the added request is the first *non-processed* request in the queue. If it is, that request/thread is responsible for triggering the GC for itself, and any subsequently added requests (until the GC runs the next round of request processing).

However, there's a snatch to this proposal. The request processing is done inside `Metaspace::purge()`, and that function is skipped if the GC didn't unload any classes. The proposed logic relies on that function to always be run when a GC is running. So, I've also changed so that the GC unconditionally calls the request processing. An alternative would be to always run the Metaspace::purge() code. That might even help return memory from temporarily allocated metaspace memory earlier, but I've left that exercise for a potential future improvement.

I've also tweaked the test so that we get a bit more info if this test fails again.

Testing: I could reliably reproduce the original hang on my macbook laptop, within a few minutes. With this fix I can run the test in a loop for hours without reproducing the hang. I've tested this together with Generational ZGC code, running tier1-tier7 on Linux x64. I've started more extensive testing on openjdk/jdk.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292697](https://bugs.openjdk.org/browse/JDK-8292697): ZGC: Hangs when almost out of metaspace memory

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [04998b3c](https://git.openjdk.org/jdk/pull/9985/files/04998b3c0c12e4343f511b6f5a66ab9b51a39659)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [aa89d763](https://git.openjdk.org/jdk/pull/9985/files/aa89d763ad221002504614bf05a2e8d66c097264)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9985/head:pull/9985` \
`$ git checkout pull/9985`

Update a local copy of the PR: \
`$ git checkout pull/9985` \
`$ git pull https://git.openjdk.org/jdk pull/9985/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9985`

View PR using the GUI difftool: \
`$ git pr show -t 9985`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9985.diff">https://git.openjdk.org/jdk/pull/9985.diff</a>

</details>

### CODE REVIEW

**Review by fisk — APPROVED** [review]
Looks good.

**Comment by stefank** [comment]
Thanks Erik!

**Review by coleenp — APPROVED** [review]
This seems ok.  The comments help describe this unexpected interaction.

File: test/hotspot/jtreg/vmTestbase/metaspace/shrink_grow/ShrinkGrowTest/ShrinkGrowTest.java [hunk 1]
```diff
@@ -145,6 +145,7 @@ private void go() {
         // step 2: try to load one more class
         // it should be impossible
         try {
+            log("just a tiny mint");
```
[the remark this answers is not in the document]
Comment (stefank) [reply test/hotspot/jtreg/vmTestbase/metaspace/shrink_grow/ShrinkGrowTest/ShrinkGrowTest.java:148]: I wanted extra logging to see if we reached this statement (we didn't). And I though this line fitted the context, similar to the other rather silly statements in the test (eatALittleMemory, "washing hands before meal"). I can change it if it bothers you.

File: src/hotspot/share/memory/metaspaceCriticalAllocation.cpp [hunk 2]
```diff
@@ -113,13 +113,47 @@ void MetaspaceCriticalAllocation::remove(MetadataAllocationRequest* request) {
...
+          return curr == request;
+        }
+      }
+
+      return false;
+    };
+
+    if (is_first_unprocessed()) {
+      // The first non-processed request takes ownership of triggering the GC
+      // on behalf of itself, and all trailing requests in the list.
```
[the remark this answers is not in the document]
Comment (stefank) [reply src/hotspot/share/memory/metaspaceCriticalAllocation.cpp:152]: Just style:
1) I wanted to pull this logic out to a function/lambda
2) I wanted to keep the code near its usage
3) I didn't want all ceremony of having to declare the function / or creating a free function taking two extra parameters.

File: test/hotspot/jtreg/vmTestbase/metaspace/shrink_grow/ShrinkGrowTest/ShrinkGrowTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dcubed-ojdk) [reply test/hotspot/jtreg/vmTestbase/metaspace/shrink_grow/ShrinkGrowTest/ShrinkGrowTest.java:148]: s/just a tiny mint/just a wafer thin mint/

https://en.wikipedia.org/wiki/Mr_Creosote

[the remark this answers is not in the document]
Comment (dcubed-ojdk) [reply test/hotspot/jtreg/vmTestbase/metaspace/shrink_grow/ShrinkGrowTest/ShrinkGrowTest.java:148]: Thanks for the laugh for my day!!

[the remark this answers is not in the document]
Comment (stefank) [reply test/hotspot/jtreg/vmTestbase/metaspace/shrink_grow/ShrinkGrowTest/ShrinkGrowTest.java:148]: Yeah, I should have looked up the exact quote. :)

File: src/hotspot/share/memory/metaspaceCriticalAllocation.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (stefank) [reply src/hotspot/share/memory/metaspaceCriticalAllocation.cpp:152]: I think code gets easier to read if you separate out the two concerns:
1) What is it you want to check
2) What is it you want to do if that happens

I find that stye of code easier to read, and easier to add precise comments to.

[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/memory/metaspaceCriticalAllocation.cpp:152]: 🤷  Fine, it's a matter of taste then.

**Comment by stefank** [comment]
Thanks for the reviews!
/integrate

File: src/hotspot/share/memory/metaspace.cpp [hunk 3]
```diff
@@ -995,23 +995,30 @@ const char* Metaspace::metadata_type_name(Metaspace::MetadataType mdtype) {
   }
 }
 
-void Metaspace::purge() {
+void Metaspace::purge(bool classes_unloaded) {
```
Comment (tstuefe) [inline src/hotspot/share/memory/metaspace.cpp:998]: Would it not have been more logical to heave both `MetaspaceCritical_lock` and `MetaspaceCriticalAllocation::process();` out of this function? As it is, `Metaspace::purge`, called with false, is now a no-op that only does things on behalf of the GC. It has nothing to do with the functioning of the allocator.

Comment (stefank) [reply src/hotspot/share/memory/metaspace.cpp:998]: I understand what you mean. On the other hand, if you take a look at Metaspace::allocate, it does quite a lot more than being an allocator, for example, it calls into the GC code (which performs the other side of the metaspace critical allocations).

I wouldn't be opposed if anyone wanted to restructure this.

**Pull request closed without merging**